app.py

```diff
@@ -856,28 +874,32 @@ def tree_process(tree : ParentedTree, senttok: List[str]) -> tuple[ParentedTree,
 	"""
 	# guardrails against producing illict tree structures
 	tree_copy = tree.copy(deep=True)
-	# if initial parse labels non-gaps as GAP, change to N-Head by default
-	# if initial parse labels punctuation as something other than punctuation, change to proper punctuation category
-	# if initial parse labels non-punctuation as punctuation, change to N-Head
+	ptree_terminals = []
 	for subt in tree_copy.subtrees(lambda t: t.height() == 2):
 		i = subt[0]
+		# if initial parse labels non-gaps as GAP, change to N-Head by default
 		if subt.label.startswith('GAP') and senttok[i] != '_.':
 			subt.label = 'N-Head'
-		# -LRB- and -RRB- are options for the grammar, so we need to escape them in case parser uses them to label terminals
-		elif subt.label == '-LRB-':
-			subt.label = 'LRB-p'
-		elif subt.label == '-RRB-':
-			subt.label = 'RRB-p'
-		# ensure that label is a punctuation label for punctuation nodes. exception: symbols that might not be punctuation.
+		# if the label is a punctuation sequence, make sure it has a function ('p' by default)
+		# (handles cases where token is changed to punctuation in the text window)
+		if is_possible_punct_token(subt.label) and subt.label == '':
+			subt.label = SYMBOL_LABEL+"-p"
+		# if the token is unambiguously punctuation, make sure it has 'p' function
 		if is_possible_punct_token(senttok[i]) and senttok[i] not in AMBIG_SYM:
-			for e in PUNCT_ESCAPING:
-				if senttok[i] == e['ptree_token']:
-					subt.label = e['ptree_label']
+			subt.label = SYMBOL_LABEL+"-p"
+		# if the token has function 'p' and has an idiosyncratic POS tag, change to that POS tag
+		if senttok[i] in np.concatenate(list(PUNCT_LABELS.values())) and is_punct_label(subt.label):
+			for label, tokens in PUNCT_LABELS.items():
+				if senttok[i] in tokens:
+					subt.label = label
 					break
```

Comment:
It looks to me like the point of `PUNCT_LABELS` is to map from token to label—but the way the data structure is organized is the reverse. If you make it `{"-LRB-": "LRB-p", "[": "LRB-p", "{": "LRB-p" ...}`, the code will be easier to understand (and faster despite listing some strings multiple times—dict lookups are faster than for loops).

Reply:
I think you are first setting `subt.label` to `"*-p"` but then changing it if there is a match in `PUNCT_LABELS`. With reorganized PUNCT_LABELS, to show more explicitly that `"*"` is the default: `subt.label = PUNCT_LABELS.get(tok, SYMBOL_LABEL+"-p")`.

Or even, if instead of PUNCT_LABELS you had PUNCT_TAGS, whose entries lacked "-p", then it could be `subt.label = PUNCT_TAGS.get(tok, SYMBOL_LABEL) + "-p"`

Reply:
(and consider renaming `SYMBOL_LABEL` to `SYMBOL_TAG`, the difference being that labels include the function part)